File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.cs
public class Solution {
    public int SubarraysDivByK(int[] nums, int k) {
        int sum = nums[0];
        int ans = 0;
        Dictionary<int,int> remainderMap = new();
        if(sum%k == 0)
        {
            ans++;
        }
        remainderMap.Add(mod(sum,k),1);
        // Console.WriteLine($"{mod(sum,k)}");
        for(int i=1; i<nums.Length; i++)
        {
            sum+=nums[i];
            // Console.WriteLine($"{mod(sum,k)} rem for {nums[i]}");
            if(mod(sum,k) == 0)
            {
                ans++;
            }
            if(remainderMap.ContainsKey(mod(sum,k)))
            {
                ans+=remainderMap[mod(sum,k)];
                remainderMap[mod(sum,k)]++;
            }
            else
            {
                remainderMap.Add(mod(sum,k),1);
            }
        }
        return ans;
    }
    private int mod(int x, int m)
    {
        return (x%m + m)%m;
    }
}

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        HashMap<Integer,Integer> map = new HashMap<>();
        int ans = 0;
        int sum = 0;
        map.put(0, 1);
        for(int num : nums)
        { 
            sum += num;
            // Reminders can not be negative in Maths but Java does not follow this, (-3)%5 in java = -3
            // but the correct answer in maths is 2 for it
            // (sum%k + k)%k will ensure correct reminders are found and not java based reminders in case of negative numbers 
            if(map.containsKey((sum%k + k)%k))
            {
                ans += map.get((sum%k + k)%k);
            }
            map.put((sum%k + k)%k, map.getOrDefault((sum%k + k)%k, 0) + 1);
        }
        return ans;
    }
}